### ISSUE TITLE
automatic generation of manpages for documentation

### ISSUE THREAD

**Issue by luminize** [machinekit/machinekit#908]
(please correct title if something more appropriate is needed)

@mhaberler I think we can remove the machinekit-man repo from the machinekit organisation. Agree?

I have merged @mhaberler branch groff-convert-make-support into @ArcEye branch [groff-convert](https://github.com/ArcEye/machinekit/tree/groff-convert) into my own branch [groff-convert](https://github.com/ArcEye/machinekit/tree/groff-convert).

a first run of `make whole-sausage` yields this [pastebin](http://pastebin.com/y1PM1GRc) which raises some warnings and errors.
I'll take these info and work on the source files to get no warnings and no errors. When I have updated some files I'll make a PR to @ArcEye (and merge PR's to my repo+branch when there's something new)

**Comment by luminize** [comment]
or would it make more sense to make PR's of commits of updated sources against machinekit:master and rebasing the local branches?

**Comment by ArcEye** [comment]
> @mhaberler I think we can remove the machinekit-man repo from the machinekit organisation. Agree?

On the contrary, if we create a branch `groff-convert` under that, we can all PR to the same MK branch rather than having copies of private repos dotted around.

* Whatever conversion tool(s) is / are used they must be available as packages to Wheezy and Jessie.
in order to meet requirements of docker or whatever containers, plus building on supported platforms.

pandoc does not do direct groff to html/xml conversion, but does lots of others.
Knowing the output format required will determine the tool probably.

I will look at the components that are throwing errors, a lot of the old .comps had troff formatting embedded in them, especially in the comment headers.

**Comment by mhaberler** [comment]
I'm trying now two formats:

docbook to html - just to see what the result looks like, but not a good base format thereafter:

	`xsltproc -o $@.html /usr/share/xml/docbook/stylesheet/docbook-xsl/xhtml-1_1/docbook.xsl $@`

and see what that yields:
also https://blogs.gnome.org/pmkovar/2015/10/27/converting-docbook-into-asciidoc/, to see if the resulting asciidoc can be carried further as a base format

**Comment by ArcEye** [comment]
All the .icomp man pages are already formatted and spaced to produce good .asciidoc pages

They are converted by outputting to a .txt file and then converting with pandoc, with will also do .md and .html
conversions

https://github.com/machinekit/machinekit-docs/blob/master/machinekit-documentation/maintainence/man2asciidoc.sh

**Comment by ArcEye** [comment]
https://github.com/ArcEye/machinekit/tree/groff-convert/man/examples

has a txt file generated directly from an icomp man page and the subsequent pandoc conversions 
of that txt file into asciidoc, md and html

**Comment by luminize** [comment]
@mhaberler @ArcEye I will continue with making sausages and correcting the component sources until I have no errors with doclifter. 

Since we aim to generate readable/renderable information, let's not get blinded by sticking with asciidoc at least for these generated manpages. Seems pointless to me since one shouldn't change the generated asciidoc anyway. So a chain with as little conversions/steps seems easiest to me.

**Comment by mhaberler** [comment]
yes, agree  - step1 is to correctly convert anything to DocBook - that is the better representation than a .txt generated by `man` 

A cursory initial conversion docbook to html can be had by adding to the rule like so: https://github.com/mhaberler/machinekit/commit/14db7d18f9cacfe70b5529af36d6c52306438819

Step 2 is the conversion to whatever we decide to carry forward as manpage source format ; I think asciidoc is better than github markdown even if it requires a formatting step; also the rest of the docs is asciidoc anyway

**Comment by ArcEye** [comment]
OK, but just be mindful that errors like this
```
"man9/bitslice.9icomp", line 27: error - ill-formed NAME section ' \fBbitslice\fR' in man9/bitslice.9icomp, giving up.
"man9/bitslice.9icomp", line 27: error - uninterpreted 'SH' command
"man9/bitslice.9icomp", line 78: there were 2 errors during translation
```
that occur when running doclifter on the man file from icomps, are failing on something which has already
been generated by another program, so the source cannot be 'corrected'

Need to find out why exactly the same formatting which works with nearly every other icomp, throws errors
with bitslice, lutn and wcompn

The comp man pages that fail are also generated, but I have not altered that formatting at all, so the error could be embedded in the .comp file itself

**Comment by luminize** [comment]
@ArcEye My feeling it's mostly with omitting some trivial info in the source files, like the string after the name in the groff source like here https://github.com/luminize/machinekit/commit/cfc46aaa3bf2610c80e0708ff1e26982d38abade

I just dumbly ran Michaels addition for `.html` and `.asciidoc` rules and there's a big difference between the output of files man/xml-docs/man9/weighted_sum.9.xml.asciidoc is one of the best yet, while man/xml-docs/man9/siggen.9.xml.asciidoc doesn't come near.
The .html generated files look OK so far.

I'll look into the differences later today. I think there is a big cleanup needed in the different source files. No shortcuts which were "ok" up to now :)

**Comment by ArcEye** [comment]
It is largely the source files - I agree.
Found 2 instances just now where `/f` is not followed by `B` or `I` and thus it swallows the first letter of the word which should be highlighted or italicised and throws a warning.

Check the ones I have already committed before going back to it later.

**Comment by luminize** [comment]
Just a moment of reflection...

Are we doing the right thing?

Shouldn't we stop using troff in the sources, and use asciidoc/markdown/reST or whatnot, and use that to generate other formats (docbook, HTML etc plus generate manpages) from that.

I mean, troff as source isn't very inviting for people to make contributions.

**Comment by mhaberler** [comment]
static manpages: yes, once converted to asciidoc, let's forget about the nroff source

now, of course if somebody comes around to generate asciidoc from comp, then... 

but if the conversion nroff/docbook/asciidoc work without manual intervention, we're good for that part too

**Comment by ArcEye** [comment]
The problem with abandoning legacy man pages, is on the minimalist, single board, non internet connected setup.

I personally don't use them and can always look up what I need on the web, but nearly all the other users seem to think it is mandatory to use just this type of low power ARM board for everything with MK.

>  now, of course if somebody comes around to generate asciidoc from comp, then... 

A challenge I shall resist for now :tongue:

**Comment by luminize** [comment]
Well, I think that we could still have manpages for those who want them, but have them generated from asciidoc in the source.

Instead of: source with troff -> manpage -> docbook/whatnot -> html 

Source with asciidoc -> manpage or docbook/whatnot or html 

I understand that this needs comp to generate a proper asciidoc. I'm wondering aloud if reviewing troff in the sources and manpages as part of the building documentation chain is a future proof setup.

The main reason: lowering the barrier. It's not really a 2016 way for documenting in troff. Big hurdle teeth grinding experience. Especially first time. Think about manpage for socfpga, or pru documentation... I would certainly be held back by troff. I did the delayline manpage and i didnt really got excited by that.

We can discuss this in a separate issue if that's more appropriate.

**Comment by ArcEye** [comment]
@luminize 
After a nights rest, I think the only sane way to go is to have one base format for all docs as you say.

This will entail a fair bit of work, but it will only have to be done once.

There is no hurry to get the end docs produced, that was simply prompted by @mhaberler's discovery of the ease of use of Jenkins for CI.

The underlying mess we have discovered converting from an unsatisfactory base format (troff / groff / nroff) with additional formatting embedded in actual sources of binaries, probably requires a root and branch approach.

1.  Convert all static man1 and man3 pages to asciidoc.  Probably just requires scripting a mass conversion and manually checking results and tweaking as required.

2.  Change comp / instcomp to generate asciidoc pages for components

3.  Change .comp and .icomp component code to alter groff `/fB... /fR` type bold highlighting embedded in the source, to asciidoc `*...*` etc.

4.  Change .c sources to generate asciidoc.  They are run through comp, so the sources again will need any groff `/fB... /fR` type bold highlighting embedded in the source, replaced with asciidoc `*...*` etc.  AFAIK there is not much or any, most of the text used is lifted from the MODULE_XXXX  module information, so those need to exist and be as meaningful as possible.  The other formatting is applied by comp and will be dealt with.

If this seems like a plan, I suppose I will have to do No. 2 :frowning: 

Once we have all the docs present or being built in asciidoc format, the makefile conversions, web hooks etc will be the easy part.

Then we probably back to the aborted document conversion discussion of last year and may yet get a CI based `machinekit.io/docs` !!

**Comment by mhaberler** [comment]
![image](https://cloud.githubusercontent.com/assets/901479/14490501/7d56e030-0173-11e6-8095-0b334bddeeef.png)

from:  http://elcubanocafe.blogspot.co.at/2011/07/hook-line-and-sinker.html

**Comment by luminize** [comment]
> On 13 Apr 2016, at 11:03, ArcEye <notifications@github.com> wrote:
> 
> @luminize <https://github.com/luminize> 
> After a nights rest, I think the only sane way to go is to have one base format for all docs as you say.
> 
> This will entail a fair bit of work, but it will only have to be done once.
> 
I agree, let’s not make rash decisions. It will cost us a lot of sweat :)
> There is no hurry to get the end docs produced, that was simply prompted by @mhaberler <https://github.com/mhaberler>'s discovery of the ease of use of Jenkins for CI.
> 
we can still decide to first get some output to publish (getting results as quickly and basic as possible). I think the html generated pages are pretty good so far. That will at least get us some continuous up-to-date basic documentation out into the open.
> The underlying mess we have discovered converting from an unsatisfactory base format (troff / groff / nroff) with additional formatting embedded in actual sources of binaries, probably requires a root and branch approach.
> 
I think that roo. Also, there’s one step before No 1. Test if the .asciidoc is good enough to use as a source. If we have trouble later on with converting asciidoc to manpage or the other formats we might have chosen the wrong format.
A bit of searching yielded me this: http://www.methods.co.nz/asciidoc/chunked/ch24.html but also asciidoctor should be able to do this http://asciidoctor.org/docs/render-documents/#selecting-an-output-format (see ‘manpage’)
> Convert all static man1 and man3 pages to asciidoc. Probably just requires scripting a mass conversion and manually checking results and tweaking as required
> 
I will look if I can find a good converted .asciidoc manpage in the sausage build and convert it back to manpage. That will give us an idea what it takes (we’ll need that for No. 2 )
> Change comp / instcomp to generate asciidoc pages for components
> 
> Change .comp and .icomp component code to alter groff /fB... /fR type bold highlighting embedded in the source, to asciidoc *...* etc.
> 
> Change .c sources to generate asciidoc. They are run through comp, so the sources again will need any groff /fB... /fR type bold highlighting embedded in the source, replaced with asciidoc *...* etc. AFAIK there is not much or any, most of the text used is lifted from the MODULE_XXXX module information, so those need to exist and be as meaningful as possible. The other formatting is applied by comp and will be dealt with.
> 
> If this seems like a plan, I suppose I will have to do No. 2 
> 
happy to help out when we decide to do this 
> Once we have all the docs present or being built in asciidoc format, the makefile conversions, web hooks etc will be the easy part.
> 
> Then we probably back to the aborted document conversion discussion of last year and may yet get a CI based machinekit.io/docs !!
>

**Comment by ArcEye** [comment]
Ignoring @mhaberler's joy at suckering us into hard work...... :facepunch: 

If you can do *1.* for now,  I am already part way through doing *2.*

When I have instcomp producing full .asciidoc man pages, you will be able to apply reverse conversion to troff and conversions to html, then we will see how well it converts and what tweaks may be required.

**Comment by luminize** [comment]
> On 13 Apr 2016, at 12:53, ArcEye <notifications@github.com> wrote:
> 
> Ignoring @mhaberler's joy at suckering us into hard work...... 
> 
> If you can do 1. for now, I am already part way through doing 2.
> 
I will do 1. :)

**Comment by RunningLight** [comment]
Just noticed this flurry of activity. Have you tried the Ruby gem `asciidoctor`? From its documentation

> Asciidoctor is a fast text processor and publishing toolchain for converting AsciiDoc content 
to HTML5, EPUB3, PDF, DocBook 5 (or 4.5) slidedecks and other formats.

Well, it looks like they lied about PDF---at least for the version I'm using, and asciidoctor-pdf is in its infancy---but I was able easily to create an asciidoc file for `comp` by hand and then generate comp.1 and comp.html files from it. 

To generate comp.html
```
$ asciidoctor -d manpage comp.adoc
```

To generate comp.1
```
$ asciidoctor -d manpage -b manpage comp.adoc
```

Have a look at https://drive.google.com/folderview?id=0BxgxhFDH9ROSYWxSX0dkN2o4SWM&usp=sharing

Note: I chose to install asciidoctor using the Ruby infrastructure to make sure I got a recent version
```
sudo gem install asciidoctor
```

This loaded asciidoctor 1.5.4 (into Ruby 1.9.3p194) on my Debian Wheezy 7.9 VM.

**Comment by RunningLight** [comment]
PS - There is an anomaly in my manpage output. The asciidoc markup for italic text was converted to underscored text (compare _sudo_ in the HTML vs the manpage).

**Comment by ArcEye** [comment]
> Have you tried the Ruby gem asciidoctor? 

Yes I have used it previously.

Presently we are getting all the man page documentation _written_ in .asciidoc

Then we will try all the different conversion programs to see which gives the best results for conversion to man page and to html, it may not be the same one!

**Comment by ArcEye** [comment]
@luminize 

I am doing *3.* now for icomps

Out of necessity, because until they are cleaned up, I cannot test the output from instcomp properly and be sure what is causing any glitches.

Thus far looking promising, instcomp is generating .asciidoc files direct into man/man9 for now

**Comment by luminize** [comment]
> On 13 Apr 2016, at 15:51, ArcEye <notifications@github.com> wrote:
> 
> @luminize
> 
> I am doing 3. now for icomps
> 
> Out of necessity, because until they are cleaned up, I cannot test the output from instcomp properly and be sure what is causing any glitches.
> 
> Thus far looking promising, instcomp is generating .asciidoc files direct into man/man9 for now
> 
@ArcEye Oh my, sounds like your up to speed. Do you have something to push to github? Then i'll pull and work on no 1 with your generated asciidocs

**Comment by ArcEye** [comment]
> Do you have something to push to github?

Just done it
https://github.com/ArcEye/machinekit/commits/groff-convert
Last 4 commits with todays date

*Current state of play*

instcomp is modified to produce .asciidoc files in man/doc/man9
Most .icomps have been edited to remove groff formatting and to improve their appearence in asciidoc

Some icomps remain to be done
bldc_hall3 for example has huge wadges of groff formatting in tables, which need converting

That's me for today :smile:

**Comment by mhaberler** [comment]
whoa! great to see the end of the seventies.. ;)

**Comment by mhaberler** [comment]
hm2_soc status:

- usability now improved - firmware can be loaded HAL-time
- boot-time fpga loading not needed
- firmware will live in the standard `/lib/firmware/...` location, using the hostmot2-built in firmware code

any number of 
`````
loadrt hostmot2
loadrt hm2_soc config="firmware=socfpga.rbf"
unloadrt hm2_soc
loadrt hm2_soc config="firmware=socfpga.rbf"
unloadrt hm2_soc
````
commands succeeds - no more crashes.

The key contribution was - thanks to @the-snowwhite's stamina: the `gpio_altera` module wreaks havoc on fpga loading.

the proper incantation (from shell) is [this](https://github.com/mhaberler/machinekit/blob/431ee9c459661744bbc242e545503b0089243e37/scripts/hm2soc_program_fpga). The hm2_soc.c driver [sticks exactly to this sequence](https://github.com/mhaberler/machinekit/blob/431ee9c459661744bbc242e545503b0089243e37/src/hal/drivers/mesa-hostmot2/hm2_soc.c#L180-L258) - all peachy.

**Comment by luminize** [comment]
I did some tinkering and the asciidoc files that got generated are a beauty to read.

I first tried generating the manpage with `asciidoctor` (the ruby gem) but I ran into troubles with the backend. I wanted to use something working out of the box, so I `sudo apt-get install asciidoc` which unfortunately pulls in tex atcetera. Not sure if we want that dependency on the beaglebones etcetera, but that's for later. I can imagine this being a problem if you build from source PLUS you'd want to generate manpages next to nicer) asciidoc files.

So after installing `asciidoc` i only had to do some minor changes to get a resulting manpage. It got generated with the following command `a2x --doctype manpage --format manpage test2_and2.asciidoc` (test2_and2 was my copy of the generated and2.asciidoc)

The source which was used (revision 1 was the generated asciidoc, revision 2 were the changes that had to be made to get the manpage to be generated by a2x):
https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767/revisions

the resulting manpage
https://gist.github.com/luminize/c660fb695cf10656fa2d43d6320e1f24

some notes:
* following conventions must be used regarding the description header (first line), first NAME section and second SYNOPSIS section http://www.methods.co.nz/asciidoc/chunked/ch24.html
* I had to unindent some headers from `=== Header` to just 2 like `== Header` but I saw that adding more indentation works too (so `=== another header` gives more indentation).
* the manpage that gets generated gets it's file name from the NAME section, so if I were to change this line https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767#file-asciidoc-to-roff-L8 in `foobar - Two-input AND gate` then the manpage would be `foobar.9`
* there's apparently no way of generating a component with other extensions than digits, like `.9icomp`
* the date when the the manpage was made seems to be in the middle of the last line https://gist.github.com/luminize/c660fb695cf10656fa2d43d6320e1f24/revisions#diff-751dc2761e629b132579370b7b9217ddR48 and in this case the `:man version:` tag too
* these 3 lines https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767#file-asciidoc-to-roff-L2-L4 seem to bring some extra info like the manual (HAL Components), source (and2.icomp) and version (date? commit?)
* these lines https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767/revisions#diff-9eab729d04b39dd001b319f419b37043L9 couldn't be used since a2x first converts to `.xml` and that step throws an error

**Comment by mhaberler** [comment]
@luminize whatever formatting is needed could possibly be shifted to build time - the asciidoc toolchain on-target certainly doesnt make sense

**Comment by luminize** [comment]
@mhaberler ok, was just more of an observation. I find it a bit hard to think that the youth of today is in very urgent need of manpages anyways, especially if one is developping.

**Comment by luminize** [comment]
@ArcEye I did some reshuffeling and small fixes and now instcomp should generate convertable asciidoc files. https://github.com/luminize/machinekit/commit/d2b0f1b6432de9515a4dc396f2e34c754f0d6420 the thing I didn't do yet is to get the info on the source file name+extension (like component.icomp) into the attributes like here https://github.com/luminize/machinekit/blob/d2b0f1b6432de9515a4dc396f2e34c754f0d6420/src/hal/utils/instcomp.g#L977
That would be nice to have the manpage show the source file

**Comment by ArcEye** [comment]
> the resulting manpage
https://gist.github.com/luminize/c660fb695cf10656fa2d43d6320e1f24

The man page looks OK but what does the asciidoc look like?
All the things taken out were there to improve the format and appearance of the asciidoc.

> these 3 lines https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767#file-asciidoc-to-roff-L2-L4 seem to bring some extra info like the manual (HAL Components), source (and2.icomp) and version (date? commit?)

They don't add anything, that info (except the source, which is obvious but could easily be added) was in the top of the page anyway.
That is how the existing instcomp man pages were, because that is how instcomp has written them for some while, so writing that way directly to man page works fine, it is just converting back from asciidoc that is causing problems.

**The answer to this is so glaringly obvious that we have all failed to see it in our concentration upon conversion to and from different formats**.

The current code produces perfectly good manpages which work in the legacy manner.

If I duplicate the writing process to **also** produce an .asciidoc for the same component, for minimal time processing, we get a custom formatted version which is best able to render good quality output in the chosen medium.

I don't have a lot of time today, but since both outputs are already written, it will just be a case of combining and creating switches to trigger the output required and building those into the Submakefile.

What do you think?

Should make things much easier - any tweaks to improve output will not have knock-on effects into the other format.

**Comment by luminize** [comment]
> On 14 Apr 2016, at 10:37, ArcEye <notifications@github.com> wrote:
> 
> the resulting manpage
> https://gist.github.com/luminize/c660fb695cf10656fa2d43d6320e1f24 <https://gist.github.com/luminize/c660fb695cf10656fa2d43d6320e1f24>
> The man page looks OK but what does the asciidoc look like?
> All the things taken out were there to improve the format and appearance of the asciidoc.
> 
mostly the things I did were to get a2x to convert asciidco ->  xml -> manpage without errors.
the resulting asciidoc:  https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767
> these 3 lines https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767#file-asciidoc-to-roff-L2-L4 <https://gist.github.com/luminize/6b63537b68c2cd506a9f28a536309767#file-asciidoc-to-roff-L2-L4> seem to bring some extra info like the manual (HAL Components), source (and2.icomp) and version (date? commit?)
> 
> They don't add anything, that info (except the source, which is obvious but could easily be added) was in the top of the page anyway.
> 
I thought that it would more clear to see if you’re dealing with an icomp or comp
> That is how the existing instcomp man pages were, because that is how instcomp has written them for some while, so writing that way directly to man page works fine, it is just converting back from asciidoc that is causing problems.
> 
correct
> The answer to this is so glaringly obvious that we have all failed to see it in our concentration upon conversion to and from different formats.
> 
> 
> The current code produces perfectly good manpages which work in the legacy manner.
> 
yes, but at least taking manpages out of the building chain for documentation has advantages.
> If I duplicate the writing process to also produce an .asciidoc for the same component, for minimal time processing, we get a custom formatted version which is best able to render good quality output in the chosen medium.
> 
> I don't have a lot of time today, but since both outputs are already written, it will just be a case of combining and creating switches to trigger the output required and building those into the Submakefile.
> 
> What do you think?
> 
smart thinking. Provided that we keep asciidoc as the way to document, and forget all about roff in the component sources. That’s really a big hurdle and downplays the ease of comp. I really like that this way asciidoc and manpage get generated without additional dependencies. Asciidoc can be used for building/converting to other formats, and the manpages should not be part of building other stuff. I think there’s not a lot maintenance on instcomp for that. And we pave the way for a way to get a modern way of source documentation with little effort.
> Should make things much easier - any tweaks to improve output will not have knock-on effects into the other format.
> 
yes. added bonus!

**Comment by ArcEye** [comment]
Good.

I am building in switches to instcomp for man page docs and asciidocs.

These can be used by the Submakefile, but also work internally if someone uses instcomp to generate docs themselves.
By selecting or not, man page builds can be completely removed from the build, we could even make it a configure switch (or cmake define) which determines the doc build and does not do so unless specifically selected.  (one for later)

> They don't add anything, that info (except the source, which is obvious but could easily be added) was in the top of the page anyway.
I thought that it would more clear to see if you’re dealing with an icomp or comp

When I get to carry the same changes over to comp, I will put something in to instcomp and comp docs to show which it is, albeit the section about how instcomps are invoked, should be a clue for those :wink:

**Comment by ArcEye** [comment]
OK, instcomp now creates both man pages and asciidocs
https://github.com/ArcEye/machinekit/commit/c1c3243a892db3ce6102237b3d7ff6da6f650bea

These are selectable using --document (-d) and --ascii-document (-a) switches
The install option will work for both types, I have created the code but not updated the switches yet.
The view option is just restricted to man page, as this is the only one that makes sense in a terminal environment.

The Submakefile currently creates both types by default - we need that to test for now.
`make clean` for asciidoc not working, need to look at it later.

I have changed some component source formatting to something that renders a reasonable page in both groff and asciidoc (and2, or2, xor2 all had asciidoc lists which did not work for groff)
The rest of the .icomps output need checking between the icomp man pages and aciidoc, but that is all I have time for until later today

Feel free to play:)

**Comment by luminize** [comment]
> On 14 Apr 2016, at 12:33, ArcEye wrote:
> 
> Feel free to play:)

Will do!

**Comment by ArcEye** [comment]
https://github.com/ArcEye/machinekit/commit/ee26eda60fda48e03052e1b0f93640ddf2bd2b6f
Fixes the make clean and inserts file origin for the docs as requested.

The --install-???? switches should now work too, not had time to test yet.

Should have more time tomorrow, so will try to sort out the source files with heavy groff formatting in tables etc. eg. bldc_hall3 and then check through all the icomp generated files in both formats.

**Comment by luminize** [comment]
> On 14 Apr 2016, at 18:50, ArcEye <notifications@github.com> wrote:
> 
> ArcEye@ee26eda
> Fixes the make clean and inserts file origin for the docs as requested.
> 
> The --install-???? switches should now work too, not had time to test yet.
> 
> Should have more time tomorrow, so will try to sort out the source files with heavy groff formatting in tables etc. eg. bldc_hall3 and then check through all the icomp generated files in both formats.
> 
Been a bit busy myself. Hope to have some more time end of tomorrow afternoon.

**Comment by ArcEye** [comment]
Starting to look at the components that have major groff formatting for tables etc.

What is actually present is a very 'clever' groff if /else block which uses one set of formatting for a html table and another for man page.

This results in man pages being rendered well and asciidoc being unrecognisable.

The tables (bldc, lut etc) are embedded in the DESCRIPTION field which is lifted as a single block by comp / instcomp from the option `description ".................";` in the component, so no easy possibility of intervention.

We either need to completely rewrite components that display perfectly as man pages, or perhaps introduce another option heading for specialised formatting, such as tables, after `description`.

Something like options of `special-format-man ".............";`  and  `special-format-doc ".............";`
Then comp / instcomp can simply read in the option field that applies and render that.

We will just need to format an asciidoc table to replace the html one then, should be able to cut and paste into a blank doc and convert with pandoc etc.

It is a bit of work for a few components, but with those as examples, it will pretty much 'future-proof' the system for any similar uses.

Unless I hear stunning alternative thoughts, that is where I am heading.

**Comment by ArcEye** [comment]
> Something like options of special-format-man "............."; and special-format-doc ".............";
Then comp / instcomp can simply read in the option field that applies and render that.

That works perfectly :+1: 

man pages exactly as before, just need to convert the html tables to asciidoc 

Commissioning new laptop interfering with progress right now!

**Comment by luminize** [comment]
> On 15 Apr 2016, at 13:57, ArcEye <notifications@github.com> wrote:
> 
> Something like options of special-format-man "............."; and special-format-doc ".............";
> Then comp / instcomp can simply read in the option field that applies and render that.
> 
> That works perfectly 
> 
> man pages exactly as before, just need to convert the html tables to asciidoc
> 
need help?
> Commissioning new laptop interfering with progress right now!
> 
that's a perfectly good reason.

**Comment by ArcEye** [comment]
> need help?

If you could update your repo and do a build, that should give you a set of man pages in /man/man9 and a set of asciidoc in /man/doc/man9

Ignore any with tables etc in them, I will be dealing with those

Need to compare the man page, asciidoc page and the source component and make any alterations required to the source to get the best output.

(eg. Noticed that if there is a blank line after or before `"""` multi-line text, a '*' charactor gets output in groff  -- that kind of thing)_

Don't alter the formatting in instcomp for now or we are not comparing like with like.
Now that we are not converting groff > asciidoc, that seems to work well.

**Comment by ArcEye** [comment]
Sitrep

Tables converted to asciidoc instead of HTML and instcomp options work perfectly in deciding which version to render.

src/hal/i_components edited from a - hbridge and committed
https://github.com/ArcEye/machinekit/commit/542b540e33aa876481465135ce2cf52f31a989b9

I'll edit in ascending alphabetical order until we meet somewhere, so keep me updated.

Enough for now, my eyes need a rest ;)

**Comment by mhaberler** [comment]
@ArcEye - I feel with you ;)

**Comment by luminize** [comment]
> On 16 Apr 2016, at 13:26, ArcEye <notifications@github.com> wrote:
> 
> Sitrep
> 
> Tables converted to asciidoc instead of HTML and instcomp options work perfectly in deciding which version to render.
> 
> src/hal/i_components edited from a - hbridge and committed
> ArcEye@542b540 <https://github.com/ArcEye/machinekit/commit/542b540e33aa876481465135ce2cf52f31a989b9>
> I'll edit in ascending alphabetical order until we meet somewhere, so keep me updated.
> 
will do, bit of a chaotic weekend so my time will be a bit low. Mostly in the evening when the noise levels in the house have died down.
> Enough for now, my eyes need a rest ;)
> 
well earned!

**Comment by ArcEye** [comment]
Next batch done 
https://github.com/ArcEye/machinekit/commit/41ff20c1a0eb410f72df1040ef511ddc03c18a4b

@luminize What did you find was the best conversion from the built in man pages to asciidoc?

Once the icomps are done, need to convert all the pre-existing xxxx.9 man pages.

**Comment by luminize** [comment]
@ArcEye 
I did some small tests with converting the manual written manpages (man9/*.9) and I'll convert all of those manpages to .asciidoc this evening with a small adapted script. They only need small manual cleanup afterwards to be able to generate the asciidoc back to manpages for those who can't live without manpages.

After that we need to check the formatting of instcomp that gets the same layout (in asciidoc) for both the manually written (which need to have the ability to be converted back to manpage) and the ones generated by instcomp. (headers and other trivial things)

**Comment by mhaberler** [comment]
just a heads up on the next step - build integration - and what is happening there:

@kinsamanka did the docker images for building machinekit on all of amd64,i386,armhf and for wheezy, jessie and raspbian - those are used by the travis builds and also the jenkins server I've set up

now we're about to extend hose images by a few packages needed for other tasks like kernel builds and uboot (https://github.com/machinekit/mk-builder/issues/2) 

rebuilding the docker images is a very slow process, so not phun to do too often - we have not started this yet as there is no urgency (I did an interim docker image which addresses my needs)

but it would be good to think through which packages you need for manpage generation eventually so we can slide these packages into the next image build

then the mk-builder scripts (jessie-64 - need to do this on one target only) can take care of any formatting and upload the results (to wherever - TBD)

I guess `asciidoc pandoc doclifter` and then what?

no urgency on this - good enough to state your shopping list when you are done

**Comment by luminize** [comment]
>but it would be good to think through which packages you need for manpage generation eventually so we can slide these packages into the next image build

>then the mk-builder scripts (jessie-64 - need to do this on one target only) can take care of any formatting and upload the results (to wherever - TBD)

>I guess asciidoc pandoc doclifter and then what?

even better, I converted the asciidoc to manpage with `asciidoctor -d manpage -b manpage classicladder.9.asciidoc` afaik `asciidoctor` is a very minimal package (other than asciidoc which needs tex and whatnot)

**Comment by mhaberler** [comment]
yeah the stock asciidoc is a tad top-heavy ;)

**Comment by ArcEye** [comment]
@mhaberler 
> I guess asciidoc pandoc doclifter and then what?
no urgency on this - good enough to state your shopping list when you are done

We should not need doclifter at all, as we are going to generate man pages from asciidoc.

I suspect just pandoc and asciidoctor, those are the only 2 that I normally use, we should know soon.

@luminize 
> After that we need to check the formatting of instcomp that gets the same layout (in asciidoc) for both the manually written (which need to have the ability to be converted back to manpage) and the ones generated by instcomp.

Rather than change instcomp to fit the formatting of handwritten man pages, which have no standardisation and no common generator, we would be better changing the far smaller number of handwritten pages to match instcomp, which we know does produce consistent rendered man pages and asciidocs.

Then we have one standard - our own - not something we inherited from what whoever wrote a page decided to do.

**Comment by luminize** [comment]
> Rather than change instcomp to fit the formatting of handwritten man pages, which have no standardisation and no common generator, we would be better changing the far smaller number of handwritten pages to match instcomp, which we know does produce consistent rendered man pages and asciidocs.
> 
> Then we have one standard - our own - not something we inherited from what whoever wrote a page decided to do.
> 
I'll check the conversion from one of the asciidoc generated from instcomp to manpage and see how much slack we have in our own sectionning. 

I think in the end the limitations (if any) from generating manual written asciidoc to manpages with asciidoctor will determine the level of required standardization.

I'll update in a few hours :)

**Comment by ArcEye** [comment]
You beat me to the post @luminize 

I have been experimenting with converting the existing legacy static man pages into asciidoc.
Have tried all sorts of combinations, doclifter into xml, pandoc xml to asciidoc etc etc.
I am not impressed by the results.  

They all require manual reformatting to look anything like as good as the asciidocs and man pages from instcomp.
Certainly not any sort of automatic process.

This is a one-off effort with the legacy man pages and the required corresponding asciidocs.  
Once they are converted, they will only ever require editing for changes.
We don't need to build these man pages or asciidocs each build
  
Everything else regards components, is generated automatically from a build.

I know the answer, but have been resisting it, hoping that an 'off-the-shelf' conversion tool would make it all easy for us.

I think I need to write a documenting GUI.
It should have all the possible fields from a standard man page and a free text one for AOB.
The basic engine already exists in instcomp.
(subject to a couple of tweaks, one @luminize  has already highlighted, using a progression of header sizes = / == / === for grumble free conversion to html )

In converting the existing man pages, all we then would need to do is cut and paste the actual text from each heading into it and it would render man page and asciidoc according to the formatting rules already established.

The same GUI would be able to be used for any new documentation of a similar nature, with the benefit
that it would all be standardised.

NB.  Conversion from asciidoc into html is another matter.
I know that works very well, it is the native conversion for asciidoctor and it comes out little changed from the asciidoc but in html format.
So conversion of asciidocs for web base documentation is not the problem

Check out the examples
https://github.com/ArcEye/machinekit/commit/10b202e156bc19f248e2d7828731d3f60433a731

**Comment by ArcEye** [comment]
Something like this

![selection_007](https://cloud.githubusercontent.com/assets/770392/14612509/45771264-058f-11e6-91cb-6d22854c7f2d.png)

**Comment by luminize** [comment]
> On 18 Apr 2016, at 17:51, ArcEye <notifications@github.com> wrote:
> 
> In converting the existing man pages, all we then would need to do is cut and paste the actual text from each heading into it and it would render man page and asciidoc according to the formatting rules already established.
> 
that's not needed in this case. I’ll have them done this evening.

**Comment by luminize** [comment]
great! Could serve as a template

do you plan to read an asciidoc back into those text area’s?

One thing we could do is make some bare template (in asciidoc format) which people can copy and extend upon.
(but I guess that the amount of new manpage writings will be a bit low)

**Comment by luminize** [comment]
@ArcEye well... not really finished this evening, but tomorrow will be better since I nog have the workflow ready. Strange that there are always small glitches that need to be worked out :)

The converted manpage -> asciidoc : https://github.com/luminize/machinekit/commit/620c16f20640ffa790033841866f4ff18e2badbe
and the fixed up page for generating for the to be automated asciidoc -> manpage : https://github.com/luminize/machinekit/commit/c9709cec1a52cacd4288cc04714ab1267071fe73

I continue tomorrow again ;)

**Comment by ArcEye** [comment]
Good work, I will plough on with the.icomp changes and output checking.

Once that is done, need to carry over the instcomp changes into comp, so that it too automatically produces both types of docs and then we are not far from the light.

**Comment by ArcEye** [comment]
instcomps finished https://github.com/ArcEye/machinekit/commit/088b764a14bf1d61880c1e9d46eae7f361d30fd2

Now amending `comp` to produce same output as `instcomp`

(That and 6 other things at once - as per usual)

**Comment by RunningLight** [comment]
You folks are doing great work.

A minor quibble about rebranding in man pages. I think it's a bit of a stretch to say, for example, "This man page written by Alex Joni, as part of the Machinekit project". I know I'm always too wordy for other peoples' taste, but it would be more accurate and less confrontational to say something like "This man page originally written by Alex Joni, as part of the LinuxCNC project, and subsequently edited for use in the Machinekit project".

**Comment by ArcEye** [comment]
I have been using similar to
"This man page written by Alex Joni and is *now* part of the Machinekit project"

@luminize has the hand written man pages in his conversions for the most part.

**Comment by ArcEye** [comment]
Wrong button!

**Comment by ArcEye** [comment]
comp now converted to produce man pages and asciidoc pages
https://github.com/ArcEye/machinekit/commit/b1ca6ee1ce1b8fde676bc1b22a774b44a015b8f4

Perhaps the most arcane and convoluted component of them all, `bldc.comp` converted to use comp new formatting as a test case.

Enough for one day ;)

**Comment by ArcEye** [comment]
Final changes to comp / instcomp and edit of remaining .comp files
https://github.com/ArcEye/machinekit/commit/c879fa96bca4a5afe54559ccc7b381f4753f1085

All generated man pages now produced with matching asciidoc page too

**Comment by luminize** [comment]
> On 20 Apr 2016, at 13:44, ArcEye <notifications@github.com> wrote:
> Final changes to comp / instcomp and edit of remaining .comp files
> ArcEye@c879fa9 <https://github.com/ArcEye/machinekit/commit/c879fa96bca4a5afe54559ccc7b381f4753f1085>
> All generated man pages now produced with matching asciidoc page too
> 
we probably should think how to handle the manual written asciidoc pages, since they also end on .asciidoc
Should we let the manually written pages end on .9.asciidoc, so can we then remove the .9comp.asciidoc and .9icomp.asciidoc
???
Is this correctly covered then?

**Comment by ArcEye** [comment]
> Should we let the manually written pages end on .9.asciidoc, so can we then remove the .9comp.asciidoc and .9icomp.asciidoc

There are no .9comp.asciidoc files or similar being generated

There are /man/man9/.9  .9comp and .9icomp  
That is a peculiarity of the man pages

The asciidocs of man9 pages are in /man/doc/man9
They all have .asciidoc extension,  it is their location which indicates the content not the extension. 

If you are producing asciidocs with convoluted extensions, I can only suggest the same policy.

**Comment by luminize** [comment]
> On 20 Apr 2016, at 13:59, ArcEye <notifications@github.com> wrote:
> 
> Should we let the manually written pages end on .9.asciidoc, so can we then remove the .9comp.asciidoc and .9icomp.asciidoc
> 
> There are no .9comp.asciidoc files or similar being generated
> 
> There are /man/man9/.9 .9comp and .9icomp
> 
> That is a peculiarity of the man pages
> 
> The asciidocs of man9 pages are in /man/doc/man9
> They all have .asciidoc extension, it is their location which indicates the content not the extension.
> 
> If you are producing asciidocs with convoluted extensions, I can only suggest the same policy
> 
ok, gotcha, I understand, so that means we just have to copy the manual written manpage from man/man*/*.asciidoc to man/doc/man*/ and use (later) man/doc/man*/ for whatever comes next, like generating HTML etc.

**Comment by ArcEye** [comment]
It is just a shorthand notation at present to indicate where the documents are from.

It would probably be much better in due course to have
man/doc/programs_and_utilities (man1)
man/doc/realtime_api  (man3)
man/doc/components  (man9)
or something similar

The average user has no idea about the division of man pages and it is only the arcane requirements of `man`
that gives rise to the man directory names.

**Comment by ArcEye** [comment]
I am going to have a break from it this afternoon.

Next job will be to write in a configure switch to build asciidoc man pages , off by default, to keep the crowd running on swallow-spit and tiny processors happy :smiling_imp:

**Comment by ArcEye** [comment]
> Next job will be to write in a configure switch to build asciidoc man pages , off by default, to keep the crowd running on swallow-spit and tiny processors happy :smiling_imp:

Just done and tested
https://github.com/ArcEye/machinekit/commit/fbb6cb6e8cd8e51499f58983284ffb01efb6e2ce

By default, no generated asciidocs.

`./configure --enable-build-documentation=asciidoc` (which is the only doc type allowed) 
builds asciidocs for all .icomp and .comp files

**Comment by luminize** [comment]
https://github.com/luminize/machinekit/commit/63a414da5555e9daa7977fc52e0fab0287b45645 converted 5 files to asciidoc. More later

**Comment by luminize** [comment]
strange: there a gladevcp(1) and a gladevcp(9).
Hasn't gladevcp become obsolete? May I delete both?

**Comment by ArcEye** [comment]
> strange: there a gladevcp(1) and a gladevcp(9).
Hasn't gladevcp become obsolete? May I delete both?

I should leave them for now - gladevcp is not as obsolete as it was hoped it would become, lots of people use it still.

I will have to study both manpages and see if they can be combined, @mhaberler did one and the other may be the default linuxcnc tool entry (man1)

**Comment by mhaberler** [comment]
@luminize ahem - only if you come up with a better sequel ;)

**Comment by ArcEye** [comment]
@luminize 
Let me know where you are with the conversions - have you got to man3 rtapi yet?  Need a hand?

Don't get hung up on converting static man pages to asciidoc, then converting that back to man pages.
We don't actually need it, since the man pages are static, all we need are consistent man pages with static asciidocs to match, so that we can build on-line documentation from them and the generated ones.

We don't need to create the static man pages as part of the build process, indeed there are good reasons not to.
Extra size, extra compilation time, extra processing power, extra dependencies etc.

I am writing a 'DocumentWizard' currently, that will create man pages and or asciidoc from a set of text fields the user inputs.
These will be identical to the ones comp and instcomp produce.

**Comment by luminize** [comment]
> On 22 Apr 2016, at 09:41, ArcEye <notifications@github.com> wrote:
> 
> @luminize 
> Let me know where you are with the conversions - have you got to man3 rtapi yet? Need a hand?
> 
I'll batch convert the manpage to asciidoc and then I'll push to my repo so it's only a matter of fixing up small things.
> Don't get hung up on converting static man pages to asciidoc, then converting that back to man pages.
> We don't actually need it, since the man pages are static, all we need are consistent man pages with static asciidocs to match, so that we can build on-line documentation from them and the generated ones.
> 
Well, I'd like one source, so the asciidocs should be the only ones to maintain and we should get rid of having to know about roff 
> We don't need to create the static man pages as part of the build process, indeed there are good reasons not to.
> Extra size, extra compilation time, extra processing power, extra dependencies etc.
> 
Very little dependencies. Only asciidoctor afaict 
> I am writing a 'DocumentWizard' currently, that will create man pages and or asciidoc from a set of text fields the user inputs.
> These will be identical to the ones comp and instcomp produce.
> 
Ok, so in order to get consistent asciidoc that still have the ability to get generated to manpage we must adhere to a few formatting rules. So that IMO means that what instcomp generates should have the same formatting for consistency (although we don't generate manpage from those)

**Comment by ArcEye** [comment]
> Well, I'd like one source, so the asciidocs should be the only ones to maintain and we should get rid of having to know about roff
Very little dependencies. Only asciidoctor afaict

Just trying to save you work :smile: 

>  to get generated to manpage we must adhere to a few formatting rules.

What are they?

My problem is that all the trials I did, did not produce man pages worth s**t direct from asciidoc, without a lot of tweaking that degraded the asciidoc formatting and appearence.
As we already have man pages that display properly, I am a little reluctant to ditch them in favour of something produced at the expense of the 'modern' display format.

The other radical option, is to convert all the existing man pages into a 'comp' type script.
This can run though a cut down comp which solely produces documentation.

Then all the man and asciidoc are generated from source at build time, no conversions and the formatting can be different for both.

**Comment by luminize** [comment]
At the moment I've done these 5 + delayline man9 pages https://github.com/luminize/machinekit/commit/63a414da5555e9daa7977fc52e0fab0287b45645
So if you commit some changed pages and add the pages you've done in the commit header it should be easy to see where we're at

the batch converted man1 and man3 and man9 pages are here:
https://github.com/luminize/machinekit/commit/88997334c1bf621597cbd372854d6fd87f6b31fc
https://github.com/luminize/machinekit/commit/7709b15d21e447e88ce3e408c0be290629986639
https://github.com/luminize/machinekit/commit/55eb3eba8f551a82f4fc06df91c586e8c2aa5fce

to see if the conversion back from asciidoc to manpage isn't broken I prefix the original manpage with an underscore and convert the asciidoc with `asciidoctor -d manpage -b manpage classicladder.9.asciidoc` for example. I then use 2 terminals, one with `man classicladder` and one with `man -l _classicladder.9` to compare.

known things to watch out for:
* [one dash instead of two](https://github.com/luminize/machinekit/commit/7709b15d21e447e88ce3e408c0be290629986639#diff-776ddec115040de0b3c38ae1f399ecb8R9)
* [i only use 1 empty line above the header](https://github.com/luminize/machinekit/commit/7709b15d21e447e88ce3e408c0be290629986639#diff-776ddec115040de0b3c38ae1f399ecb8R19)
* linuxcnc references [like here](https://github.com/luminize/machinekit/commit/55eb3eba8f551a82f4fc06df91c586e8c2aa5fce#diff-122246cf39f528e2db721955262cc694R109)
* [some roff formatting that hasn't been corrected](https://github.com/luminize/machinekit/commit/7709b15d21e447e88ce3e408c0be290629986639#diff-776ddec115040de0b3c38ae1f399ecb8R21) that's why I also view the generated (back from asciidoc) manpage.
* Another reason I view the generated manpage is that the lines don't get justified, so the only thing to see if line breaks go haywire is to view.

**Comment by luminize** [comment]
> On 22 Apr 2016, at 10:49, ArcEye <notifications@github.com> wrote:
> 
> Well, I'd like one source, so the asciidocs should be the only ones to maintain and we should get rid of having to know about roff
> Very little dependencies. Only asciidoctor afaict
> 
> Just trying to save you work 
> 
thx!
> to get generated to manpage we must adhere to a few formatting rules.
> 
> What are they?
> 
ah, the messages just crossed each other
apart from the (9) i generated at the first header (just now I’m thinking that the generated files I mentioned above should have (3) in the header and (1).
So hold on a bit and I’ll correct those pages with my script
> My problem is that all the trials I did, did not produce man pages worth s**t direct from asciidoc, without a lot of tweaking that degraded the asciidoc formatting and appearance.
> 
yes, that’s where my script from 1 year ago proved some help.
> As we already have man pages that display properly, I am a little reluctant to ditch them in favour of something produced at the expense of the 'modern' display format.
> 
> The other radical option, is to convert all the existing man pages into a 'comp' type script.
> This can run though a cut down comp which solely produces documentation.
> 
> Then all the man and asciidoc are generated from source at build time, no conversions and the formatting can be different for both.
> 
I'm not sure if this is less work. There still need to be eyes checking the result. It’s a lot of work, especially some really large files.
We can do those one by one,
first looking into the asciidoc if it’s viewable (directly usable for docs on website etc (automatic generation))
keep the legacy roff written manpages around until everything is cleaned up in the asciidoc files
Later remove the legacy files so we can generate manpages from the asciidocs (for this who want to)

**Comment by luminize** [comment]
ok, one step further

@ArcEye I've committed all the raw converted manpage to asciidoc (https://github.com/luminize/machinekit/tree/groff-convert), with now the correct type in the first header. like the `9`, `1`, `3`, `3hal`, `3hm2` and `3rtapi`

```source,asciidoc
= manpage(3hm2)
```
This is needed for generating a manpage from the asciidoc

Further more, a lot of times (legacy manpages), after a pin has been explained, there is an indented section where some more explanation is given about that pin. Like an example manpage:
```
PINS
       delayline.enable bit in

           Enables the delayline, when not enabled,
           the input sample is set directly to the
           output pin.
```

The only way to get that rendered with asciidoctor, while retaining the markup like bold, italic etc, as well as correctly converting with an indentation back to the manpage format is to put the text inside a block, and specify the indentation. like so: https://github.com/luminize/machinekit/commit/52956bc9fbe150329e56bd54a90e9752f796e424#diff-fb20b7fe835f3f246f48c0006257a7f0R60

**Comment by luminize** [comment]
@mhaberler @ArcEye 
could you render/view [this asciidoc page](https://raw.githubusercontent.com/luminize/machinekit/52956bc9fbe150329e56bd54a90e9752f796e424/man/man9/delayline.9.asciidoc)?
I hope we can make this the format/layout that we need for the future (so we know where to work to). It renders to a html with the following: `asciidoctor delayline.9.asciidoc` [with this result](https://gist.github.com/luminize/1896557b8334c9057b5b6f1435cd6562#file-delayline-9-html)
(copy raw to local file and open with browser)

If we're agreed I can take a look as what i'd need to change in my script to not have @ArcEye and me do a lot less manual viewing.....

**Comment by ArcEye** [comment]
>  It renders to a html with the following: asciidoctor delayline.9.asciidoc with this result
(copy raw to local file and open with browser)

That didn't actually work, but I copied the asciidoc and converted locally with `asciidoctor test.asciidoc` and it renders fine as html.

Big question, what does the man page converted back from it look like?

**Comment by luminize** [comment]
>Big question, what does the man page converted back from it look like?

https://gist.github.com/luminize/48c85f16d00349b8cc938ebfc683208f
(the gist is without the bold/italic formatting :( )

**Comment by luminize** [comment]
we might run into trouble with time limits for travis.
not sure if this is due to instcomp
https://travis-ci.org/luminize/machinekit/jobs/124999960#L2000

**Comment by luminize** [comment]
nope, must be due to some makefile additions I did in the Submakefile.
please forget my comment

**Comment by ArcEye** [comment]
> https://gist.github.com/luminize/48c85f16d00349b8cc938ebfc683208f
(the gist is without the bold/italic formatting :( )

Looks fine

> nope, must be due to some makefile additions I did in the Submakefile.
please forget my comment

I was just about to riposte that you haven't enabled asciidoc builds, so it can't be 
I won't bother now. :)

**Comment by luminize** [comment]
OMG, only up to one third of hostmot2 https://github.com/luminize/machinekit/commit/126a90dda003d73bf6e95a4082b8e831e890f745

will continue tomorrow.

Let's face it, this will need manual work to get it into shape and future proof.

**Comment by ArcEye** [comment]
Just surfaced again, out from under the worst of the flu, and looked at your repo @luminize 

I am seriously concerned that we are doing the wrong thing and landing ourselves with weeks of pointless work.

If an automatic conversion from man page to asciidoc, results in a document that requires such heavy manual editing so that it can be used to generate a man page back again - why are we doing it?
We already have the man page.

The conversion displays fine as an asciidoc and html page.
The whole point of conversion tools is labour saving.

If we replace groff formatted man pages with asciidoc pages that are so unintuitively formatted, for the sole purpose of producing groff man pages back again, we are just replacing one problem with another.

We are requiring anyone in the future producing documentation, to do it in a completely prescriptive manner, any deviation from which will result in an improperly rendered man page.

Might as well just insist they do the document in groff and have done with it.

Why don't we either:
* just get rid of groff manpages completely
* leave the existing ones where they are , with minimal rebranding etc. and make 21st century versions in asciidoc available

Whatever there will be html versions available on line

**Comment by luminize** [comment]
> On 25 Apr 2016, at 10:19, ArcEye <notifications@github.com> wrote:
> 
> Just surfaced again, out from under the worst of the flu, and looked at your repo @luminize
> 
> I am seriously concerned that we are doing the wrong thing and landing ourselves with weeks of pointless work.
> 
I understand your concern. But i think the main reason for the work imo is the different authors of the roff pages which make converting to asciidoc hard.
> If an automatic conversion from man page to asciidoc, results in a document that requires such heavy manual editing so that it can be used to generate a man page back again - why are we doing it?
> We already have the man page
> 
Imo we're doing it to get a source in a human readable and understandable format. Asciidoc as base, and not roff
> The conversion displays fine as an asciidoc and html page.
> The whole point of conversion tools is labour saving.
> 
If you have a better solution/tool, i'm all for it.
> If we replace groff formatted man pages with asciidoc pages that are so unintuitively formatted, for the sole purpose of producing groff man pages back again, we are just replacing one problem with another.
> 
I dont think the asciidoc pages are unintuitively formatted.
> We are requiring anyone in the future producing documentation, to do it in a completely prescriptive manner, any deviation from which will result in an improperly rendered man page.
> 
Correct.
> Might as well just insist they do the document in groff and have done with it.
> 
> Why don't we either:
> 
> just get rid of groff manpages completely
I'm not against that.
> leave the existing ones where they are , with minimal rebranding etc. and make 21st century versions in asciidoc available
I wouldn't choose 2 sources: confusing
> Whatever there will be html versions available on line
>

**Comment by ArcEye** [comment]
I am not sure what para you are replying to in all cases. :confused: 

The only advantage of man pages, is display in a terminal

Rendering of all "man pages" from .asciidoc, could be done simply by a variation of a script that I wrote
(only dependencies asciidoctor and lynx)

```
# 'mank' script
# usage mank component
# !/bin/bash

asciidoctor $(MANDOCPATH)/$1.asciidoc -o - | lynx -stdin
```
There are only 3 components I know of that it does not render properly, because of how asciidoctor writes html tables
So that would be 3 to fix instead of all of them.

**Comment by cdsteinkuehler** [comment]
Stupid newbie question:  Where is the current hostmot2 documentation?

I have committed to Michael H. that I will (finally) craft a man page for the hal_pru_generic driver.  I grabbed the latest from github:machinekit/machinekit, and the hostmot2 man page I want to use as a starting point still looks like groff/troff/whatever, while I was expecting asciidoc.  What's the procedure for adding a fairly major new man page?  Is there a converted version of the hostmot2 manpage available to use as a starting point?

**Comment by luminize** [comment]
https://github.com/luminize/machinekit/commit/768da341589e16aa4a727d9ccfd2eba3c8bae7e5 is my latest commit (from yesterday in the train) that has (albeit we're questioning the manual work) hostmot2 asciidoc.

**Comment by luminize** [comment]
@mhaberler @ArcEye @cdsteinkuehler 

I think we have wandered off the path, in all our enthusiasm we've made this from an issue into a new project.
The problem I have is the `linux-i-want-to-get-something-done-but-first-must-learn-more-tools-dependency-phenomenon-and-i-don-t-know-how-long-that-will-take-and-of-im-succesfull-or-need-to-go-find-another-tool`. 
As opposed to know that if I continue as I do now, it might not be smart, but I'm confident I'll finish it.

What I mean by that is in order to automatically convert the current manpages (as opposed to manually patching the current results up) is to go on a quest to make doclifter first convert to asciidoc,check results, go on subquests for tweaks and then to check if we can convert back (manpage: use asciidoc as future manpage source) and forward (html/whatnot as nice rendered output).

That quest might take me personally 1 day, or it might take me 10. I don't know.

So I think we should hack up this problem in many more small problems (as we not tend to do, things tend to grow in our enthusiasm).

this problem IMO falls apart into 3 parts:

1. get the changes of instcomp into mainline to have readable/renderable component manpages in `man/man*/`
2. manpage -> asciidoc
 - (2.a) go on the doclifter manpage to asciidoc quest to have as good as possible automatically converted asciidoc sources 
 - **OR** 
 - (2.b) keep the 80% converted asciidocs as we have them now, and try to lever the community to patch them up. Get them involved in the PR workflow, without having to be a programmer per se, get some involvement?A real hand on exercise that doesn't waste time.
3. once we have proper asciidocs we'll make the automatic build of asciidoc to html.
4. get the html into something to serve it.
5. and make an automated build to manpage again for whatever reasons.

My preference would be to do the 2b option. Not because it's technically smart, but to see if we can get people to buy in into help maintaining the project without having to go on personal quests in search for knowledge. That should be something we should do more often. 
- We grow involvement and a small skill set for community members.
- Making small pieces that are doable by more people. Instead of an "hello world" bogus introduction to some needed skill (git fear for example) people will learn them and see that the fixes are actually used in the project. That's very satisfactory.

Ideas/additions/remarks?

**Comment by ArcEye** [comment]
* 1 instcomp and comp will now produce man pages and asciidocs for all components built with them.
That is finished, makefiles amended and configure altered so that `--enable-build-documentation=asciidoc`
passed to configure, triggers asciidoc generation as well as man pages.
It could be pushed tomorrow.

* 2 All component asciidocs will generate into good readable html pages just using
`asciidoctor nameofcomponent.asciidoc`

* 3 We already have man pages for all the legacy C components / utilities / tools /GUIs
They just need some minor rebranding.

* 4 Bas's conversion means we now have asciidocs for all those same legacy C binaries, which can be converted into html pages as per 2 above.

The sticking point is an obsession with being able to convert the converted asciidocs from the legacy C binaries **back** into groff man pages again.

**We don't need to, we already have the man pages**

The bigger question in this age of broadband, mobile wifi etc. should be, whether we ditch groff man pages altogether.

If you need terminal access, as I mentioned above, you can render an asciidoc page to terminal quite simply with
`asciidoctor docname.asciidoc -o - | lynx -stdin `

I have a script which does just that.

Afraid I think throwing it open to 'the community' to help do something which IMHO does not need doing in the first place, is just procrastination.

(No offence intended @luminize, just my frustration at the state of affairs spilling out)

**Comment by luminize** [comment]
well, I thought manpages should be kept around because all people wanted this. I personally don't mind ditching the manpages. Reading asciidoc source is fine IMO.

but I won't maintain asciidoc + legacy roff manpages. That's the thing that will be confusing, and I simply think thats wrong. Period. So _either_ asciidoc needs to be able to be converted back, _or_ the legacy manpages must be amputated (chirugically be removed). No way in-between.

If we do that, and forget the roff manpages altogether, then we can get all into machinekit master and focus on the automating of asciidoc -> html -> served/rendered somewhere on each PR

**Comment by ArcEye** [comment]
I think it needs a vote, or at least a sounding of opinions.

I am going to put something on the forum, will send you a draft first.

**Comment by luminize** [comment]
> On 27 Apr 2016, at 12:39, ArcEye <notifications@github.com> wrote:
> 
> I think it needs a vote, or at least a sounding of opinions.
> 
Yes, good thought!
> I am going to put something on the forum, will send you a draft first.
>

**Comment by mhaberler** [comment]
I am all for keep _all_ docs in asciidoc, as it is defacto docbook which retains full structural information and a good base for translation, eg to hmtl

just to understand the issue with back conversion to manpage format (ok to call me an idiot), but assume we have all manpages in asciidoc - where is the showstopper converting asciidoc to man like  (works fine for delayline.9.asciidoc and kins.9.asciidoc for example) just for the packaged manpages (not for any web delivery)?

````
asciidoc -d manpage -b docbook -okins.9.xml  kins.9.asciidoc
xsltproc  /usr/share/xml/docbook/stylesheet/docbook-xsl/manpages/docbook.xsl kins.9.xml
man -l kins.9
````

(stolen from czmq)

**Comment by luminize** [comment]
There isn't any real showstopper, except manual touchup (work) on the asciidco source (that got generated from the manpages in the first place)
your example is 99% the same output as `asciidoctor -d manpage -b manpage kins.9.asciidoc`

the things re formatting back into manpage i ran into:
* first header must gave (9) in name, like `= component(9)` or `= component(3hal)` where the stuff between `( )` will give the component its filename
* there needs to be a header `== ` with the name like so
```
== NAME
kins - kinematics definitions for LinuxCNC
```
* further more to indent an explanatory note on pin descriptions one must use (taken from hostmot2) an indent to get readable manpage section with rendered (bold, underline) formatting. A code block will just be rendered as code, all formatting gone
```
=== Pins:

(float, in) hm2___<BoardType>__.__<BoardNum>__.dpll.NN.timer-us

[indent=4]
====
This pin sets the triggering offset of the associated timer. There are 4 timers
numbered 01 to 04, represented by the NN digits in the pin name. 
The units are micro-seconds. Negative numbers indicate that the trigger should
occur prior to the main hostmot2 write. It is anticipated that this value will
be calculated from the known bit-count and data-rate of the functions to be
triggered. Alternatively you can just keep making the number more negative
until the over-run error bit in the encoder goes false. 
The default value is set to 100uS, enough time for approximately 50 bits to be
transmitted at 500kHz. For very critical systems it may be worth reducing this 
until errors appear, and for very long bit-length or slow encoders it will need
to be increased.
====
```

**Comment by luminize** [comment]
Is there agreement for closing this issue?

**Comment by mhaberler** [comment]
YESS!!! amen & halleluja!
